### ISSUE TITLE
[competitor-port] Alternative Data Integration & Feature Engineering Pipeline: fracdiff, tsfresh, featuretools, Universe Selection, FeatureStore (QuantConnect + mlfinlab + tsfresh + featuretools patterns)

### ISSUE THREAD

**Issue by KCNyu** [KCNyu/clawock#1166]
## 竞品做法
**QuantConnect Lean** — `Data/alternative/` + `Algorithm.Framework/Selection/`（源码：https://github.com/QuantConnect/Lean/tree/master/Data/alternative、https://github.com/QuantConnect/Lean/tree/master/Algorithm.Framework/Selection）

### Alternative Data Sources (Lean)
| 数据源 | 类型 | 用途 |
|--------|------|------|
| `alphastreams/portfoliostate` | 机构持仓流 | 跟踪聪明钱 |
| `estimize/consensus` + `estimate` + `release` | 盈利预测共识 | 事件驱动交易 |
| `trading-economics` | 宏观经济指标 | 宏观择时 |
| `interest-rate` | 利率曲线 | 固收/利差策略 |
| `sec` | SEC 文档/XBRL | 基本面/事件 |

### Universe Selection Models (Lean)
- `FineFundamentalUniverseSelectionModel`：基本面筛选（市值/PE/ROE/营收增长等 100+ 指标）
- `CoarseFundamentalUniverseSelectionModel`：粗筛（美元成交量/价格/基本面）
- `ETFConstituentsUniverseSelectionModel`：ETF 成分股复制
- `QC500UniverseSelectionModel`：流动性 + 基本面筛选的 500 股票池
- `Sector/Theme ETF Universes`：行业/主题轮动

**mlfinlab** — `mlfinlab/features/fracdiff.py`（源码：https://github.com/hudson-and-thames/mlfinlab/tree/master/mlfinlab/features）
- `FractionalDifferentiation.get_weights()` / `frac_diff()`：分数阶差分（保留记忆的平稳化）
- `get_weights_ffd()` / `frac_diff_ffd()`：固定窗口分数阶差分（高效版本）
- `plot_min_ffd()`：ADF 检验寻找最小 d 值

**tsfresh** — `tsfresh/feature_extraction/feature_calculators.py` + `feature_extraction/settings.py`（源码：https://github.com/blue-yonder/tsfresh/tree/master/tsfresh/feature_extraction）
- 60+ 类时间序列特征计算器：统计量、自相关、周期性、分形维数、熵、峰值等
- `ComprehensiveFCParameters()` / `MinimalFCParameters()` / `EfficientFCParameters()`：预设参数集
- `extract_features()`：批量提取 + 并行化
- `select_features()`：基于显著性/相关性的特征筛选

**featuretools** — 自动化特征工程（DFS 深度特征合成）
- `EntitySet` + `Relationship`：多表关联建模
- `dfs()`：深度特征合成，自动生成聚合/变换特征
- `encode_features()` / `normalize_features()`：特征编码/归一化

## 我们现在怎么做的
`src/clawock/market_data/factors.py` — **仅有 9 个预注册因子**，硬编码计算逻辑
```python
RAW_FACTORS = (
    'residual_mom_1m', 'residual_mom_3m', 'residual_mom_6m',
    'relative_strength', 'breadth', 'liquidity', 'low_volatility',
    'drawdown_resilience', 'quality_profitability',
)
```
- 因子计算分散在 `factors.py` 各函数，**无统一特征管线、无版本管理、无血缘追踪**
- `src/clawock/market_data/factors.py:476-493` — 加权合成 `composite_score`，**无特征选择、无重要性排序、无冗余检测**
- `src/clawock/market_data/peer_residuals.py` / `peer_discovery.py` — 同业残差，**仅单一维度，无替代数据融合**
- `src/clawock/market_data/macro.py` — 宏观快照，**仅 5 个指标，无高频/替代宏观数据**
- `src/clawock/market_data/sentiment.py` — 情绪，**仅简单关键词匹配，无 NLP/实体/事件提取**
- `src/clawock/market_data/fund_flows.py` / `fundamentals.py` — 资金流/基本面，**仅基础字段，无深度基本面/另类数据**

## 差在哪
**机制缺失**：
1. **无替代数据标准化接入层** — 无统一 `AlternativeDataSource` 抽象、无数据血缘/版本/质量元数据
2. **无特征管线** — 无 `FeatureStore`、无特征版本/血缘/文档、无特征服务（在线/离线一致）
3. **无自动化特征工程** — 无分数阶差分、无 tsfresh 批量提取、无 featuretools DFS、无特征选择/筛选
4. **无替代数据源集成** — 无机构持仓/盈利预测/卫星/信用卡/供应链/ESG/新闻实体事件
5. **无 Universe Selection 抽象** — 无基本面/流动性/主题/ETF 成分股等多维选股池
6. **无特征质量监控** — 无 PSI/漂移检测、无特征重要性衰减追踪、无自动回滚

## 改法
新建/改动文件（≤600 LOC，纯离线核心逻辑，无新增外部服务、无新增 cron）：

1. **新建 `src/clawock/features/alt_data.py`**（~120 LOC）
   - `AlternativeDataSource` Protocol：`fetch(symbols, start, end) -> pd.DataFrame`、`metadata() -> dict`
   - 内置实现：`EstimizeSource`、`AlphaStreamsSource`、`TradingEconomicsSource`、`SecFilingsSource`、`SatelliteSource`（占位）、`CreditCardSource`（占位）、`ESGSource`（占位）
   - `AlternativeDataRegistry`：注册/发现/元数据管理

2. **新建 `src/clawock/features/universe.py`**（~100 LOC）
   - `UniverseSelectionModel` Protocol：`select(date, config) -> List[str]`
   - `FundamentalUniverse`：基本面筛选（市值/PE/PB/ROE/营收增长/负债/现金流）
   - `LiquidityUniverse`：流动性筛选（美元成交量/换手率/价格/点差）
   - `ETFConstituentsUniverse`：ETF 成分股复制（SPY/QQQ/XLK/XLF 等）
   - `ThemeUniverse`：主题筛选（AI/新能源/半导体/生物医药等关键词/行业映射）
   - `CompositeUniverse`：多模型组合（并集/交集/加权投票）

3. **新建 `src/clawock/features/fracdiff.py`**（~80 LOC）
   - `FractionalDifferentiation.get_weights(d, size)` / `frac_diff(series, d, thresh)`：扩展窗口
   - `get_weights_ffd(d, thresh, lim)` / `frac_diff_ffd(series, d, thresh)`：固定窗口
   - `find_min_d(series, max_d=1.0, step=0.01)`：ADF 检验自动寻优 d

4. **新建 `src/clawock/features/auto_features.py`**（~120 LOC）
   - `TSFreshExtractor`：封装 `tsfresh.extract_features()` + 并行配置
   - `FeatureToolsDFS`：封装 `featuretools.dfs()` + 实体集定义
   - `FeatureSelector`：`select_features()` 基于显著性/相关性/互信息
   - `FeaturePipeline`：`fit_transform(prices, alt_data) -> feature_matrix` 端到端管线

5. **新建 `src/clawock/features/store.py`**（~80 LOC）
   - `FeatureStore` 类：离线/在线一致的特征读写
   - `register_feature(name, func, version, dependencies, description, owner)`：元数据注册
   - `get_feature_vector(symbols, date, feature_names) -> pd.DataFrame`：点在时间查询
   - `compute_feature_importance_decay()`：重要性衰减追踪
   - `detect_feature_drift(psi_threshold=0.2)`：PSI 漂移检测

6. **新建 `src/clawock/features/alt_data_sources.py`**（~100 LOC，占位实现可后续填充）
   - `SatelliteImagerySource`：停车场/港口/油罐/作物
   - `CreditCardTransactionSource`：消费/零售/旅游高频
   - `SupplyChainSource`：进出口/发货/库存
   - `ESGSource`：碳排/多样性/治理评分
   - `NewsEntityEventSource`：实体/事件/情绪/新颖性

7. **新建 `config/features.json`**（~40 LOC）
```json
{
  "fracdiff": {"default_d": 0.5, "thresh": 1e-5, "auto_find_d": true},
  "tsfresh": {"parameter_set": "efficient", "n_jobs": 4, "chunksize": 100},
  "featuretools": {"max_depth": 2, "primitives": ["sum", "mean", "std", "max", "min", "count"]},
  "feature_selection": {"method": "significance", "p_value_threshold": 0.05, "correlation_threshold": 0.9},
  "universe": {
    "default": "fundamental",
    "fundamental": {"min_market_cap": 1e9, "min_roe": 0.1, "max_pe": 50},
    "liquidity": {"min_dollar_volume": 1e6, "min_price": 5}
  },
  "feature_store": {"psi_threshold": 0.2, "importance_decay_window": 63}
}
```

8. **改 `src/clawock/market_data/factors.py`**（~80 LOC 增量）
   - 重构为 `FeaturePipeline` 调用，保留 `RAW_FACTORS` 作为内置基础因子
   - 新增 `--feature-pipeline` CLI 参数，输出特征矩阵 + 元数据
   - 复用 `fracdiff`、`tsfresh`、`featuretools` 生成增强因子
   - 输出特征重要性、PSI、血缘到 run card

**规模量级**：~640 LOC 新增/改动，纯 Python，依赖 `tsfresh`/`featuretools`/`fracdiff`（可选依赖，`pyproject.toml` 需添加 `extras_require`）。

## 验收
今天会失败、改完会通过的 pytest 断言（贴在测试文件 `tests/test_features.py`）：

```python
def test_fractional_differentiation():
    from clawock.features.fracdiff import FractionalDifferentiation
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    prices = pd.Series(100 * np.exp(np.cumsum(np.random.randn(500) * 0.01)),
                       index=pd.date_range("2023-01-01", periods=500, freq="B"))
    # d=0.5 分数阶差分
    diffed = FractionalDifferentiation.frac_diff_ffd(prices, diff_amt=0.5, thresh=1e-5)
    assert isinstance(diffed, pd.Series)
    assert len(diffed) == len(prices)
    # 平稳性检验
    from statsmodels.tsa.stattools import adfuller
    adf_stat, p_val, *_ = adfuller(diffed.dropna())
    assert p_val < 0.05  # 应平稳

def test_find_min_d():
    from clawock.features.fracdiff import FractionalDifferentiation
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    # 非平稳序列：随机游走
    prices = pd.Series(100 + np.cumsum(np.random.randn(300) * 0.5),
                       index=pd.date_range("2023-01-01", periods=300, freq="B"))
    d = FractionalDifferentiation.find_min_d(prices, max_d=1.0, step=0.1)
    assert 0 < d <= 1.0

def test_universe_selection_fundamental():
    from clawock.features.universe import FundamentalUniverse
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    # 模拟基本面数据
    n = 100
    fundamental_data = pd.DataFrame({
        "ticker": [f"STK{i:03d}" for i in range(n)],
        "market_cap": np.random.lognormal(20, 1, n),
        "pe_ratio": np.random.uniform(5, 50, n),
        "roe": np.random.normal(0.15, 0.05, n),
        "revenue_growth": np.random.normal(0.1, 0.05, n),
        "debt_to_equity": np.random.uniform(0, 2, n),
    })
    universe = FundamentalUniverse(
        min_market_cap=1e9, max_pe=30, min_roe=0.1, min_revenue_growth=0.05
    )
    selected = universe.select("2023-01-01", fundamental_data)
    assert isinstance(selected, list)
    assert all(t in fundamental_data["ticker"].values for t in selected)
    assert len(selected) <= n

def test_universe_selection_liquidity():
    from clawock.features.universe import LiquidityUniverse
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    n = 100
    price_data = pd.DataFrame({
        "ticker": [f"STK{i:03d}" for i in range(n)],
        "price": np.random.uniform(10, 500, n),
        "volume": np.random.lognormal(12, 1, n),
        "dollar_volume": np.random.lognormal(15, 1, n),
    })
    universe = LiquidityUniverse(min_dollar_volume=1e6, min_price=5.0)
    selected = universe.select("2023-01-01", price_data)
    assert all(price_data[price_data["ticker"] == t]["dollar_volume"].values[0] >= 1e6 for t in selected)
    assert all(price_data[price_data["ticker"] == t]["price"].values[0] >= 5.0 for t in selected)

def test_feature_store_register_and_query():
    from clawock.features.store import FeatureStore
    import pandas as pd
    import numpy as np
    store = FeatureStore()
    def my_feature(prices): return prices.pct_change().rolling(20).mean()
    store.register_feature("mom_20", my_feature, version=1, dependencies=["close"],
                           description="20-day momentum", owner="test")
    # 查询
    prices = pd.Series(100 * np.exp(np.cumsum(np.random.randn(100) * 0.01)),
                       index=pd.date_range("2023-01-01", periods=100, freq="B"))
    vec = store.get_feature_vector(["AAPL"], "2023-01-30", ["mom_20"])
    assert isinstance(vec, pd.DataFrame)
    assert "mom_20" in vec.columns

def test_feature_drift_detection():
    from clawock.features.store import FeatureStore
    import pandas as pd
    import numpy as np
    store = FeatureStore()
    # 训练期特征分布
    train_feat = pd.Series(np.random.normal(0, 1, 1000))
    # 测试期特征分布发生漂移
    test_feat = pd.Series(np.random.normal(0.5, 1.2, 1000))
    psi = store._psi(train_feat, test_feat)
    assert psi > 0.2  # 应检测到漂移

def test_alternative_data_registry():
    from clawock.features.alt_data import AlternativeDataRegistry, EstimizeSource
    registry = AlternativeDataRegistry()
    registry.register("estimize", EstimizeSource())
    assert "estimize" in registry.list_sources()
    source = registry.get("estimize")
    assert hasattr(source, "fetch")
    assert hasattr(source, "metadata")

def test_feature_pipeline_integration():
    from clawock.features.auto_features import FeaturePipeline
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    prices = pd.DataFrame({
        f"STK{i:03d}": 100 * np.exp(np.cumsum(np.random.randn(200) * 0.01))
        for i in range(10)
    }, index=pd.date_range("2023-01-01", periods=200, freq="B"))
    pipeline = FeaturePipeline(use_fracdiff=True, use_tsfresh=False, use_featuretools=False)
    # 这里只测试接口存在
    assert hasattr(pipeline, "fit_transform")
    assert hasattr(pipeline, "get_feature_names")
```

## ROI 与风险
- **影响面**：为所有下游模型（因子、ML、LLM、规则）提供**工业级特征供给**；**某个已发布的数字会更准**（分数阶差分保留记忆、自动化特征挖掘替代手工因子、替代数据带来信息增量）；接入 #1158 AlphaModel、`CompositeAlphaModel` 做特征条件信号
- **高压线检查**：
  - live 每日流水线模型输出契约：**不碰** — 仅离线特征管线，`add_alpha_walkforward` 可选接入
  - 预注册研究契约 (`peer_residuals.load_rule_config`)：**不碰** — 现有因子保持不变，作为内置基础因子保留
  - dashboard 200KB payload 闸：**不碰** — 特征矩阵仅在评估/训练内部
- **风险**：`tsfresh`/`featuretools` 依赖重（可做可选依赖/extra）；特征数量爆炸需筛选；替代数据源多为商业/需授权（占位实现后续对接）；特征漂移监控需历史积累

**Comment by KCNyu** [comment]
Closing: this one is not just out of scope, it points the wrong way for this book's actual failure mode.

**The measured problem here is selection bias, not feature scarcity.** The published `signal_panel` selection numbers on `decision_map.json` read PBO **0.81 at t1**, 0.31 at t5, 0.10 at t20 over 70 splits — i.e. at the shortest horizon the best-looking signal is already more likely than not to be the artifact of choosing among the ones we have. Automatic feature generation (tsfresh / featuretools style) multiplies the candidate count by orders of magnitude against the same ~84 sessions of cross-section. That does not add alpha, it adds trials, and the honest gate we already run would then reject nearly everything at greater cost.

**Nine hand-registered factors is a deliberate shape, not an oversight.** Input-side evolution in this product is gated: a factor's constitution and weights may evolve, but only through explicit artifacts that a `git diff` can show and a forward-return panel can score. A feature store with lineage and automated importance ranking is the runtime-evolution shape this product deliberately does not take.

`fracdiff` and PIT handling are the parts with real merit; both are already tracked on the in-scope side by #1142 (expression engine + cross-sectional processors + PIT), which is where any of this should land if it lands. Closing this one as the adjacent superset.
